File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/22-UnionFind/LeetCode/LeetCode-1559.cpp
/// Disjoint set class
class DS {
    vector<int> sets;
public:
    DS(int n){sets.resize(n, -1);}
    
    // Find set with path compression
    int find(int u){
        if (sets[u] < 0) return u;
        
        int s = find(sets[u]);
        sets[u] = s;
        return s;
    } // end-findSet
    
    void join(int u, int v){
        if (sets[u] < sets[v]){
            sets[u] += sets[v];
            sets[v] = u;
        } else {
            sets[v] += sets[u];
            sets[u] = v;
        } //end-else
    } // end-join
    
    // Assume that "u" is a valid set head
    int size(int u){return -sets[u];}
}; // end-DS

class Solution {
public:
    bool containsCycle(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();

        DS ds(m*n);
        
        for (int i=0; i<m; i++){
            for (int j=0; j<n; j++){
                // Look at two neighbors: Left, Up
                vector<pair<int, int>> neigh = {{i, j-1}, {i-1, j}};
                
                for (auto [nr, nc]: neigh){
                    if (nr >= 0 && nc >= 0 && grid[nr][nc] == grid[i][j]){
                        int u = i*n+j;
                        int v = nr*n+nc;
                        int set1 = ds.find(u);
                        int set2 = ds.find(v);
                        if (set1 == set2){
                            if (ds.size(set1) >= 4) return true;
                        } else {
                            ds.join(set1, set2);
                        } // end-else
                    } // end-if
                } // end-for
            } // end-for
        } // end-for
        
        return false;
    } // end-containsCycle
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/22-UnionFind/LeetCode/LeetCode-1559.py
# Implement a Union/Find DS
# Find set with path compression
class DS:
    def __init__(self, n):
        self.sets = [-1]*n

    def find(self, u):
        if (self.sets[u] < 0): return u
        
        s = self.find(self.sets[u])
        self.sets[u] = s
        return s
    
    def join(self, u, v):
        if (self.sets[u] < self.sets[v]):
            self.sets[u] += self.sets[v]
            self.sets[v] = u
        else:
            self.sets[v] += self.sets[u]
            self.sets[u] = v

    # Assumes that "u" is a valid set head
    def size(self, u):
        return -self.sets[u]  
    
class Solution:
    def containsCycle(self, grid: List[List[str]]) -> bool:
        m = len(grid)
        n = len(grid[0])

        ds = DS(m*n)
        
        for i in range(m):
            for j in range(n):
                # Look at two neighbors: Left, Up
                neigh = ((i, j-1), (i-1, j))
                
                for nr, nc in neigh:
                    if (nr >= 0 and nc >= 0 and grid[nr][nc] == grid[i][j]):
                        u = i*n+j
                        v = nr*n+nc
                        set1 = ds.find(u)
                        set2 = ds.find(v)
                        if (set1 == set2):
                            if (ds.size(set1) >= 4): return True
                        else:
                            ds.join(set1, set2)
        return False
